src/mbgl/util/geometry_within.cpp

```diff
@@ -9,6 +9,14 @@ bool rayIntersect(const Point<double>& p, const Point<double>& p1, const Point<d
     return ((p1.y > p.y) != (p2.y > p.y)) && (p.x < (p2.x - p1.x) * (p.y - p1.y) / (p2.y - p1.y) + p1.x);
 }
 
+bool onBoundary(const Point<double>& p, const Point<double>& p1, const Point<double>& p2) {
+    const auto x1 = p.x - p1.x;
+    const auto y1 = p.y - p1.y;
+    const auto x2 = p.x - p2.x;
+    const auto y2 = p.y - p2.y;
+    return (x1 * y2 - x2 * y1 == 0) && (x1 * x2 <= 0) && (y1 * y2 <= 0);
```

Comment:
could you pls give more details (maybe add a comment) where this logic came from?

Reply:
`x1 * y2 - x2 * y1 == 0` is equivalent to more readable `x1 * y2 == x2 * y1`

Reply:
It represents cross product, I will add comments for this function.